# HW3B - Pandas Fundamentals

See Canvas for details on how to complete and submit this assignment.

## Introduction

This assignment transitions you from NumPy's numerical array operations to Pandas' powerful tabular data manipulation. While NumPy excels at homogeneous numerical arrays, Pandas is designed for the heterogeneous, labeled data that characterizes most real-world datasets—mixing dates, categories, numbers, and text within the same table.

You'll work with real bike share data from Chicago's Divvy system to answer questions about urban transportation patterns. Through three progressively complex problems—exploring usage patterns, analyzing rider behavior, and conducting temporal analysis—you'll discover why Pandas has become the standard tool for data analysis in Python.

The assignment emphasizes Pandas' design philosophy: named column access, explicit indexing methods (loc/iloc), handling missing data, and method chaining for readable data pipelines. You'll also see how Pandas builds on NumPy while adding the structure and convenience needed for practical data science work.

This assignment should take 3-5 hours to complete.

Before submitting, ensure your notebook:

- Runs completely with "Kernel → Restart & Run All"
- Includes thoughtful responses to all interpretation questions
- Uses clear variable names and follows good coding practices
- Shows your work (don't just print final answers)

### Learning Objectives

By completing this assignment, you will be able to:

1. **Construct and manipulate Pandas data structures**
   - Create DataFrames from dictionaries and CSV files
   - Distinguish between Series and DataFrame objects
   - Set and reset index structures appropriately
   - Understand when operations return views vs copies
2. **Apply explicit indexing paradigms**
   - Use `loc[]` for label-based data access
   - Use `iloc[]` for position-based data access
   - Access columns using bracket notation
   - Explain when each indexing method is appropriate
3. **Diagnose and explore datasets systematically**
   - Use `info()`, `describe()`, `head()`, and `dtypes` to understand data structure
   - Identify missing values with `isna()` and `notna()`
   - Calculate summary statistics across different axes
   - Interpret value distributions with `value_counts()`
4. **Filter data with boolean indexing and queries**
   - Combine multiple conditions with `&`, `|`, and `~` operators
   - Use `isin()` for membership testing
   - Apply `query()` for readable complex filters
   - Understand how index alignment affects operations
5. **Work with datetime data**
   - Parse dates during CSV loading
   - Extract temporal components with the `.dt` accessor
   - Filter data by date ranges
   - Create time-based derived features
6. **Connect Pandas patterns to data analysis workflows**
   - Formulate questions that data can answer
   - Choose appropriate methods for different analysis tasks
   - Interpret results in domain context
   - Recognize when vectorized operations outperform apply()

### Generative AI Allowance

You may use GenAI tools for brainstorming, explanations, and code sketches if you disclose it, understand it, and validate it. Your submission must represent your own work and you are solely responsible for its correctness.

### Scoring

Total of 90 points available, will be graded out of 80. Scores of >100% are allowed.

Distribution:

- Tasks: 48 pts
- Interpretation: 32 pts
- Reflection: 10 pts

Points by Problem:

- Problem 1: 3 tasks, 10 pts
- Problem 2: 4 tasks, 14 pts
- Problem 3: 4 tasks, 14 pts
- Problem 4: 3 tasks, 10 pts

Interpretation Questions:

- Problem 1: 3 questions, 8 pts
- Problem 2: 4 questions, 8 pts
- Problem 3: 3 questions, 8 pts
- Problem 4: 3 questions, 8 pts

Graduate differentiation: poor follow-up responses will result in up to a 5pt deduction for that problem.

## Dataset: Chicago Divvy Bike Share

The dataset you will analyze is based on real trip information from Divvy, Chicago's bike share system. It contains individual trips with start/end times, station information, and rider type.

Dataset homepage: https://divvybikes.com/system-data

Each trip includes:

- Trip start and end times (datetime)
- Start and end station names and IDs
- Rider type (member vs casual)
- Bike type (classic, electric, or docked)

Chicago's Department of Transportation uses this data to optimize station placement, understand usage patterns, and improve service. You'll explore similar questions that real transportation analysts investigate.

## Problems

### Problem 1: Creating DataFrames from Scratch

Before loading data from files, you need to understand how Pandas structures are built. In this problem, you'll create Series and DataFrames manually using Python's built-in data structures. This is a quick warmup to establish the fundamentals.

#### Task 1a: Create a Series

Create a Series called `temperatures` representing daily high temperatures for a week:

- Monday: 72°F
- Tuesday: 75°F  
- Wednesday: 68°F
- Thursday: 71°F
- Friday: 73°F

Use the day names as the index. Print the Series and its data type.

##### Your Code

In [1]:
import pandas as pd

# create temperature series 
temperatures = pd.Series( 
    [72, 75, 68, 71, 73],
    index=["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"]
)

# printing results 
print(temperatures) 
print ("\nData type: ", type(temperatures))



Monday       72
Tuesday      75
Wednesday    68
Thursday     71
Friday       73
dtype: int64

Data type:  <class 'pandas.core.series.Series'>


#### Task 1b: Create a DataFrame from a Dictionary

Create a DataFrame called `products` with the following data:

| product | price | quantity |
|---------|-------|----------|
| Widget  | 19.99 | 100 |
| Gadget  | 24.99 | 75 |
| Doohickey | 12.49 | 150 |

Use a dictionary where keys are column names and values are lists. Print the DataFrame and report its shape.

##### Your Code

In [2]:
# Creating the DataFrame
products= pd.DataFrame({
    "product": ['Widget', 'Gadget', 'Doohickey'],
    "price": [19.99, 24.99, 12.49],
    "quantity":[100, 75, 150]
})

# Show the data types of each column
print("Column types:\n", products.dtypes)

# Show shape 
print("\nShape:", products.shape) 

# Display DataFrame
print("\nProducts DataFrame:\n", products, "\n")

Column types:
 product      object
price       float64
quantity      int64
dtype: object

Shape: (3, 3)

Products DataFrame:
      product  price  quantity
0     Widget  19.99       100
1     Gadget  24.99        75
2  Doohickey  12.49       150 



#### Task 1c: Access DataFrame Elements

Using the `products` DataFrame from Task 1b, extract and print:

1. The `price` column as a Series
2. The `product` and `quantity` columns as a DataFrame (using a list of column names)

##### Your Code

In [3]:
# Extract price column as a Series
price_series = products['price']
print("Price column as Series:\n", price_series, "\n")
print("\nprice_series type", type(price_series))
print("\n 1d-series:", price_series.shape)
# Using a list of column names
Columns_DataFrame = products[['product', 'quantity']]
print("\nProduct and Quantity columns DataFrame:\n", Columns_DataFrame)
print("\nColumns_DataFrame type", type(Columns_DataFrame))
print("\n 2d-shape:", Columns_DataFrame.shape)

Price column as Series:
 0    19.99
1    24.99
2    12.49
Name: price, dtype: float64 


price_series type <class 'pandas.core.series.Series'>

 1d-series: (3,)

Product and Quantity columns DataFrame:
      product  quantity
0     Widget       100
1     Gadget        75
2  Doohickey       150

Columns_DataFrame type <class 'pandas.core.frame.DataFrame'>

 2d-shape: (3, 2)


#### Interpretation

Answer the following questions (briefly / concisely) in the markdown cell below:

1. Data structure mapping: When you create a DataFrame from a dictionary (like in Task 1b), what do the dictionary keys become? What do the values become?
2. Bracket notation: Why does `df['price']` return a Series, but `df[['price']]` return a DataFrame? What's the difference in what you're asking for?
3. Index purpose: In Task 1a, you used day names as the index instead of default numbers (0, 1, 2...). When would a custom index like this be more useful than the default numeric index?

##### Your Answers

***
##### Data structure mapping
A DataFrame constructed from a dictionary uses the dictionary's keys for column names, and the dictionary's values (in list or array form) are the values inside those columns.

##### Bracket notation
df['price'] returns a Series because a single column got selected, unlike when df[['price']] returns a DataFrame because a column list got selected. Single brackets signify 1D Series. Double brackets signify a 2D DataFrame.

##### Index purpose
A custom index (for example, day names) is more useful when it has meaningful labels that make the data easier to interpret or retrieve, for example, using dates, categories, or identifiers instead of generic row numbers.
***


### Problem 2: Loading and Initial Exploration

Before starting this problem, make sure you are working in a copy of this file in the `my_repo` folder you created in HW2a. You must also have a copy of the file `202410-divvy-tripdata-100k.csv` in a subdirectory called `data`. That file structure is illustrated below.

```text
~/insy6500/my_repo
└── homework
    ├── data
    │   └── 202410-divvy-tripdata-100k.csv
    └── hw3b.ipynb
```

#### Task 2a: Load and Understand Raw Data

Start by loading the data "as-is" to get a general understanding of the overall structure and how Pandas interprets it by default.

Note on file paths: The provided code uses `Path` from Python's `pathlib` module to handle file paths. Path objects work consistently across operating systems (Windows uses backslashes `\`, Mac/Linux use forward slashes `/`), automatically using the correct separator for your system. The provided code defines `csv_path` which should be used as the filename in your `pd.read_csv` to load the data file.

1. Use `pd.read_csv` to load `csv_path` (provided below) without specifying any other arguments. Assign it to the variable `df_raw`.
2. Use the methods we described in class to explore the shape, structure, types, etc. of the data. In particular, consider which columns represent dates or categories.
3. Note the amount of memory used by the dataset. See the section on memory diagnostics in notebook 07a for appropriate code snippets using `memory_usage`.

##### Your Code

In [4]:
import pandas as pd
import numpy as np
from pathlib import Path

# create a OS-independent pointer to the csv file created by Setup
csv_path = Path('./data/202410-divvy-tripdata-100k.csv')

# load and explore the data below (create additional code / markdown cells as necessary)

# # Load the data as-is 
df_raw = pd.read_csv(csv_path)

print("Shape:", df_raw.shape)

# Check data types explicitly ( We can see this in info)
print("\n---------dtype---------")
display(df_raw.dtypes)

print("\n--- df_raw.info() ---")
df_raw.info()

print ("\n-----------Quick statistical information--------")
display(df_raw.describe())

# Preview first few rows
print("\n-------------First 5 rows ----------------")
display(df_raw.head())

# Memory usage per column and total (same format as Task 7a)
print("\nMemory usage per column:")
print(df_raw.memory_usage(deep=True))

print("\nTotal memory usage:")
print(f"{df_raw.memory_usage(deep=True).sum() / 1024**2:.2f} MB")


Shape: (100000, 13)

---------dtype---------


ride_id                object
rideable_type          object
started_at             object
ended_at               object
start_station_name     object
start_station_id       object
end_station_name       object
end_station_id         object
start_lat             float64
start_lng             float64
end_lat               float64
end_lng               float64
member_casual          object
dtype: object


--- df_raw.info() ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   ride_id             100000 non-null  object 
 1   rideable_type       100000 non-null  object 
 2   started_at          100000 non-null  object 
 3   ended_at            100000 non-null  object 
 4   start_station_name  89623 non-null   object 
 5   start_station_id    89623 non-null   object 
 6   end_station_name    89485 non-null   object 
 7   end_station_id      89485 non-null   object 
 8   start_lat           100000 non-null  float64
 9   start_lng           100000 non-null  float64
 10  end_lat             99913 non-null   float64
 11  end_lng             99913 non-null   float64
 12  member_casual       100000 non-null  object 
dtypes: float64(4), object(9)
memory usage: 9.9+ MB

-----------Quick statistical information--------


start_lat      start_lng       end_lat       end_lng
count  100000.000000  100000.000000  99913.000000  99913.000000
mean       41.898817     -87.644839     41.899246    -87.645279
std         0.045897       0.027118      0.046581      0.028055
min        41.648501     -87.840000     41.610000    -89.120000
25%        41.879356     -87.658902     41.879472    -87.659172
50%        41.894666     -87.641180     41.894822    -87.641255
75%        41.925566     -87.627716     41.925858    -87.628579
max        42.070000     -87.530000     43.930000    -86.050000


-------------First 5 rows ----------------


ride_id  rideable_type               started_at  \
0  67BB74BD7667BAB7  electric_bike  2024-09-30 23:12:01.622   
1  5AF1AC3BA86ED58C  electric_bike  2024-09-30 23:19:25.409   
2  7961DD2FC1280CDC   classic_bike  2024-09-30 23:32:24.672   
3  2E16892DEEF4CC19   classic_bike  2024-09-30 23:42:11.207   
4  AAF0220F819BEE01  electric_bike  2024-09-30 23:49:25.380   

                  ended_at         start_station_name start_station_id  \
0  2024-10-01 00:20:00.674     Oakley Ave & Touhy Ave           bdd4c3   
1  2024-10-01 00:42:09.933                        NaN              NaN   
2  2024-10-01 00:23:18.647     St. Clair St & Erie St           9c619a   
3  2024-10-01 00:10:16.831  Ashland Ave & Chicago Ave           72a04d   
4  2024-10-01 00:06:27.476          900 W Harrison St           11da85   

           end_station_name end_station_id  start_lat  start_lng    end_lat  \
0                       NaN            NaN  42.012342 -87.688243  41.970000   
1    Benson Ave & Church St         a10cf0  42.070000 -87.730000  42.048214   
2  LaSalle St & Illinois St         fbd1ad  41.894345 -87.622798  41.890762   
3    Loomis St & Archer Ave         896337  41.895954 -87.667728  41.841633   
4         900 W Harrison St         11da85  41.874754 -87.649807  41.874754   

     end_lng member_casual  
0 -87.650000        casual  
1 -87.683485        casual  
2 -87.631697        member  
3 -87.657435        casual  
4 -87.649807        member


Memory usage per column:
Index                     132
ride_id               6500000
rideable_type         6150493
started_at            7200000
ended_at              7200000
start_station_name    6832669
start_station_id      5261329
end_station_name      6830739
end_station_id        5258155
start_lat              800000
start_lng              800000
end_lat                800000
end_lng                800000
member_casual         5500000
dtype: int64

Total memory usage:
57.16 MB


#### Task 2b: Reload with Proper Data Types

1. Repeat step 2a.1 to reload the data. Use the `dtype` and `parse_dates` arguments to properly assign categorical and date types. Assign the result to the variable name `rides`.
2. After loading, use `rides.info()` to confirm the type changes.
3. Use `memory_usage` to compare the resulting size with that from step 2a.3.

##### Your Code

In [5]:
import pandas as pd
import numpy as np
from pathlib import Path

# Define the data path
csv_path = Path('./data/202410-divvy-tripdata-100k.csv')

# Reload the data with correct dtypes (I included the float columns as well, even though it wasn’t necessary)
rides = pd.read_csv(
    csv_path,
    dtype={
        "ride_id": "string",
        "rideable_type": "string",
        "start_station_name": "string",
        "start_station_id": "string",
        "end_station_name": "string",
        "end_station_id": "string",
        "member_casual": "string",
        "start_lat": "float64",
        "start_lng": "float64",
        "end_lat": "float64",
        "end_lng": "float64"
    },
    parse_dates=["started_at", "ended_at"]
)

# --- Verify the changes ---
print("\n--- rides.info() ---")
rides.info()

# --- Memory usage per column and total ---
print("\nMemory usage per column:")
print(rides.memory_usage(deep=True))

print("\nTotal memory usage:")
print(f"{rides.memory_usage(deep=True).sum() / 1024**2:.2f} MB")



--- rides.info() ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             100000 non-null  string        
 1   rideable_type       100000 non-null  string        
 2   started_at          100000 non-null  datetime64[ns]
 3   ended_at            100000 non-null  datetime64[ns]
 4   start_station_name  89623 non-null   string        
 5   start_station_id    89623 non-null   string        
 6   end_station_name    89485 non-null   string        
 7   end_station_id      89485 non-null   string        
 8   start_lat           100000 non-null  float64       
 9   start_lng           100000 non-null  float64       
 10  end_lat             99913 non-null   float64       
 11  end_lng             99913 non-null   float64       
 12  member_casual       100000 non-null  string        
dtypes: datet

#### Task 2c: Explore Structure and Missing Data

Using the `rides` DataFrame from Task 2b:

1. Determine the range of starting dates in the dataframe using the `min` and `max` methods.
2. Count the number of missing values in each column. See the section of the same name in lecture 06b.
3. Convert the Series from step 2 to a DataFrame using `.to_frame(name='count')`, then add a column called 'percentage' that calculates the percentage of missing values for each column.

##### Your Code

In [6]:
# rides DataFrame from Task 2b

# date ranges 
start_date_min = rides["started_at"].min()
start_date_max = rides["started_at"].max()
print(f"Start date range:\nFrom {start_date_min} to {start_date_max}\n")

# number of missing values from lecture 06b
missing_counts = rides.isna().sum()
print("Missing values per column:")
print(missing_counts)

# Convert the Series to a DataFrame and calculate the percentage ( missing value in each column/total value in each column)
missing_df = (missing_counts.to_frame(name='count'))
percentage_value = (missing_counts / len(rides)) * 100
missing_df['percentage'] = percentage_value
print("\nMissing values per column and percentage:")
display(missing_df)





Start date range:
From 2024-09-30 23:12:01.622000 to 2024-10-31 23:54:02.851000

Missing values per column:
ride_id                   0
rideable_type             0
started_at                0
ended_at                  0
start_station_name    10377
start_station_id      10377
end_station_name      10515
end_station_id        10515
start_lat                 0
start_lng                 0
end_lat                  87
end_lng                  87
member_casual             0
dtype: int64

Missing values per column and percentage:


count  percentage
ride_id                 0       0.000
rideable_type           0       0.000
started_at              0       0.000
ended_at                0       0.000
start_station_name  10377      10.377
start_station_id    10377      10.377
end_station_name    10515      10.515
end_station_id      10515      10.515
start_lat               0       0.000
start_lng               0       0.000
end_lat                87       0.087
end_lng                87       0.087
member_casual           0       0.000

#### Task 2d: Create Trip Duration Column and Set Index

Before setting the index, create a derived column for trip duration:

1. Calculate trip_duration_min by subtracting `started_at` from `ended_at`, then converting to minutes using `.dt.total_seconds() / 60`
3. Display basic statistics (min, max, mean) for the new column using `.describe()`
4. Show the first few rows with `started_at`, `ended_at`, and `trip_duration_min` to verify the calculation
5. Set `started_at` as the DataFrame's index. Verify the change by printing the index and displaying the first few rows.

##### Your Code

In [7]:
# trip duration in minutes
rides['trip_duration_min'] = (rides['ended_at'] - rides['started_at']).dt.total_seconds() / 60

# basic statistics for the new column (.describe)
print("Trip duration statistics (in minutes):")
print(rides['trip_duration_min'].describe())

# the first few rows (started_at, ended_at and trip_duration_min)(verification)
print("\nSample rows with trip duration:")
display(rides[['started_at', 'ended_at', 'trip_duration_min']].head())

#started_at as the DataFrame index and verify
rides = rides.set_index('started_at')

print("\n'started_at' as an index verification:")
print(rides.index.name)
display(rides.head())


Trip duration statistics (in minutes):
count    100000.000000
mean         16.144576
std          52.922539
min           0.006533
25%           5.489271
50%           9.423592
75%          16.407171
max        1499.949717
Name: trip_duration_min, dtype: float64

Sample rows with trip duration:


started_at                ended_at  trip_duration_min
0 2024-09-30 23:12:01.622 2024-10-01 00:20:00.674          67.984200
1 2024-09-30 23:19:25.409 2024-10-01 00:42:09.933          82.742067
2 2024-09-30 23:32:24.672 2024-10-01 00:23:18.647          50.899583
3 2024-09-30 23:42:11.207 2024-10-01 00:10:16.831          28.093733
4 2024-09-30 23:49:25.380 2024-10-01 00:06:27.476          17.034933


'started_at' as an index verification:
started_at


ride_id  rideable_type  \
started_at                                                 
2024-09-30 23:12:01.622  67BB74BD7667BAB7  electric_bike   
2024-09-30 23:19:25.409  5AF1AC3BA86ED58C  electric_bike   
2024-09-30 23:32:24.672  7961DD2FC1280CDC   classic_bike   
2024-09-30 23:42:11.207  2E16892DEEF4CC19   classic_bike   
2024-09-30 23:49:25.380  AAF0220F819BEE01  electric_bike   

                                       ended_at         start_station_name  \
started_at                                                                   
2024-09-30 23:12:01.622 2024-10-01 00:20:00.674     Oakley Ave & Touhy Ave   
2024-09-30 23:19:25.409 2024-10-01 00:42:09.933                       <NA>   
2024-09-30 23:32:24.672 2024-10-01 00:23:18.647     St. Clair St & Erie St   
2024-09-30 23:42:11.207 2024-10-01 00:10:16.831  Ashland Ave & Chicago Ave   
2024-09-30 23:49:25.380 2024-10-01 00:06:27.476          900 W Harrison St   

                        start_station_id          end_station_name  \
started_at                                                           
2024-09-30 23:12:01.622           bdd4c3                      <NA>   
2024-09-30 23:19:25.409             <NA>    Benson Ave & Church St   
2024-09-30 23:32:24.672           9c619a  LaSalle St & Illinois St   
2024-09-30 23:42:11.207           72a04d    Loomis St & Archer Ave   
2024-09-30 23:49:25.380           11da85         900 W Harrison St   

                        end_station_id  start_lat  start_lng    end_lat  \
started_at                                                                
2024-09-30 23:12:01.622           <NA>  42.012342 -87.688243  41.970000   
2024-09-30 23:19:25.409         a10cf0  42.070000 -87.730000  42.048214   
2024-09-30 23:32:24.672         fbd1ad  41.894345 -87.622798  41.890762   
2024-09-30 23:42:11.207         896337  41.895954 -87.667728  41.841633   
2024-09-30 23:49:25.380         11da85  41.874754 -87.649807  41.874754   

                           end_lng member_casual  trip_duration_min  
started_at                                                           
2024-09-30 23:12:01.622 -87.650000        casual          67.984200  
2024-09-30 23:19:25.409 -87.683485        casual          82.742067  
2024-09-30 23:32:24.672 -87.631697        member          50.899583  
2024-09-30 23:42:11.207 -87.657435        casual          28.093733  
2024-09-30 23:49:25.380 -87.649807        member          17.034933

#### Interpretation

Reflect on problem 2 and answer (briefly / concisely) the following questions:

1. What types did Pandas assign to `started_at` and `member_casual` in Task 2a? Why might these defaults be problematic?
2. Look at the values in the station ID fields. Based on what you learned about git commit IDs in HW3a, how do you think the station IDs were derived?
3. Explain in your own words what method chaining is, what `df.isna().sum()` does and how it works.
4. Assume you found ~10% missing values in station columns but ~0% in coordinates. What might explain this? How might you handle the affected rows?

##### Your Answers

***
#### What types did Pandas assign to started_at and member_casual in Task 2a? Why might these defaults be problematic?
Pandas selected or set started_at to an `object type` instead of a `datetime`, and member_casual to an `object` type instead of a `category`. This can be a problem because objects consume more memory and don’t support time-based or categorical operations, such as sorting by date or grouping by user type, efficiently.

Note: I explicitly defined started_at as a datetime and member_casual as a string to ensure proper cleaner operations and type handling  

#### Look at the values in the station ID fields. Based on what you learned about git commit IDs in HW3a, how do you think the station IDs were derived?  
The station ID values seem to be random alphanumeric strings (for instance, "bdd4c3"), like git commit IDs. This indicates they were generated using a hashing or encoding method to make each station uniquely identifiable rather than using simple numbers.

#### Explain in your own words what method chaining is, what df.isna().sum() does and how it works. 
Method chaining means applying multiple pandas methods one after another in a single line, where each method’s output becomes the input for the next.
`"df.isna().sum()"` first checks which values are missing (`isna() `), then counts how many missing values there are per column (`sum()`).

#### Assume you found ~10% missing values in station columns but ~0% in coordinates. What might explain this? How might you handle the affected rows?
If the station columns contain around 10% missing values and the coordinates columns are almost filled in, then the location data was most likely recorded, but the station names or IDs were not linked properly. I could handle these missing rows by either removing them if they’re few, or filling them using nearby coordinates or matching patterns from similar values.

***


#### Follow-Up (Graduate Students Only)

Compare memory usage results in 2a.3 and 2b.3. What caused the change? Why are these numbers different from what is reported at the bottom of `df.info()`? Which should you use if data size is a concern?

Working with DataFrames typically requires 5-10x the dataset size in available RAM. On a system with 16GB, assuming about 30% overhead from the operating system and other programs, what range of dataset sizes would be safely manageable? Calculate using both 5x (optimistic) and 10x (conservative) multipliers, then explain which you'd recommend for reliable work.

##### Your Answers

***
#### Compare memory usage results
In Task 2a.3, the DataFrame used the default data types assigned by Pandas (object for datetimes and object for text).On the other hand, in Task 2b.3, I reloaded the dataset using explicit data types, converting text columns to string and parsing dates with datetime64. This change lowered memory usage because string and datetime store data more efficiently than generic object types.

#### `df.info` and what should I use
The numbers from `df.memory_usage(deep=True)` differ from those shown at the bottom of df.info() because df.info() gives only a rough estimate and, by default, does not include the full memory footprint of Python objects. 

What should you use?   
When memory efficiency matters, the more accurate metric is `df.memory_usage(deep=True).sum()` since it accounts for all underlying data.

#### Working with DataFrames
A DataFrame normally requires in-memory 5 to 10 times its size for intermediate results and computations. 16 GB of RAM exists in a computer. The operating system and other applications use 30% of the RAM. Python has about 11.2 GB of free memory available for use.  

Based on that:   

Optimistic (5 ×) = 11.2 GB /5 = 2.24 GB

Conservative (10 ×) = 11.2 GB /10 = 1.12 GB

Therefore, a dataset between 1 GB to 2 GB is safely manageable on a 16 GB system. For reliable work, I would recommend following the conservative limit to avoid crashes during heavy processing.
***

### Problem 3: Filtering and Transformation

With clean data loaded, you can now filter and transform it to answer specific questions. This problem focuses on Pandas' powerful indexing and filtering capabilities, along with creating derived columns that enable deeper analysis.

You'll continue working with the `rides` DataFrame from Problem 2, which has `started_at` set as the index.

#### Task 3a: Boolean Indexing and Membership Testing

Use boolean indexing and the `isin()` method to answer these questions:

1. How many trips were taken by *members* using *electric bikes*? Use `&` to combine conditions.
2. What percentage of all trips does this represent?
3. How many trips started at any of these three stations: "Streeter Dr & Grand Ave", "DuSable Lake Shore Dr & Monroe St", or "Kingsbury St & Kinzie St"? Use `isin()`.

Note: Remember to use parentheses around each condition when combining with `&`.

##### Your Code

In [8]:
# Question 1 (How many trips..) 
member_electric = rides[(rides['member_casual'] == 'member') & (rides['rideable_type'] == 'electric_bike')]
num_member_electric = len(member_electric)
print(f"Number of trips by members using electric bikes: {num_member_electric:,}")

# Question 2 ( What percentage ..)
total_trips = len(rides)
percent_member_electric = (num_member_electric / total_trips) * 100
print(f"\nPercentage of total trips: {percent_member_electric:.2f}%")

# Question 3 ( How many trips started at..) 
stations = [
    "Streeter Dr & Grand Ave",
    "DuSable Lake Shore Dr & Monroe St",
    "Kingsbury St & Kinzie St"
]
station_trips = rides[rides['start_station_name'].isin(stations)]
num_station_trips = len(station_trips)
print(f"\nNumber of trips started at the three stations: {num_station_trips:,}")

Number of trips by members using electric bikes: 33,121

Percentage of total trips: 33.12%

Number of trips started at the three stations: 2,702


#### Task 3b: Create Derived Columns from Datetime

Add two categorical columns to the rides DataFrame based on trip start time:

1. `is_weekend`: Boolean column that is True for Saturday/Sunday trips. Use .dt.dayofweek on the index (Monday=0, Sunday=6).
2. `time_of_day`: String categories based on start hour:
   - "Morning Rush" if hour is 7, 8, or 9
   - "Evening Rush" if hour is 16, 17, or 18
   - "Midday" for all other hours

For step 2, initialize the column to "Midday", then use .loc[mask, 'time_of_day'] with boolean masks to assign rush hour categories. Extract hour using .dt.hour on the index.

After creating both columns, use value_counts() on time_of_day to show the distribution.

##### Your Code

In [9]:
# Task 3b 
# Question 1 ( is_weekend)
rides['is_weekend'] = rides.index.dayofweek.isin([5, 6])

# Question 2( time_of_day)
rides['time_of_day'] = 'Midday'

# Extract hours
hrs = rides.index.hour

# Boolean masks
morning_mask = hrs.isin([7, 8, 9])
evening_mask = hrs.isin([16, 17, 18])

# Assign categories
rides.loc[morning_mask, 'time_of_day'] = 'Morning Rush'
rides.loc[evening_mask, 'time_of_day'] = 'Evening Rush'

# Store as category ( to save memory)
rides['time_of_day'] = rides['time_of_day'].astype('category')

# Show distribution
print("time_of_day distribution:")
print(rides['time_of_day'].value_counts())



time_of_day distribution:
time_of_day
Midday          55912
Evening Rush    28218
Morning Rush    15870
Name: count, dtype: int64


#### Task 3c: Complex Filtering with query()

Use the `query()` method to find trips that meet **all** of these criteria:
- Casual riders (not members)
- Weekend trips  
- Duration greater than 20 minutes
- Electric bikes

Report:
1. How many trips match these criteria?
2. What percentage of all trips do they represent?
3. What is the average duration of these trips?

Hint: Column names work directly in `query()` strings. Combine conditions with `and`.

##### Your Code

In [10]:
# Filter rides based on all the given criteria using query()
filtered_rides = rides.query(
    "member_casual == 'casual' and is_weekend == True and trip_duration_min > 20 and rideable_type == 'electric_bike'"
)

# how many trips match
num_trips = len(filtered_rides)

# What percentage of all trips do they represent
percent_trips = (num_trips / len(rides)) * 100

# the average duration of these trips
avg_duration = filtered_rides['trip_duration_min'].mean()

# Display results
print(f"Number of trips matching criteria: {num_trips}")
print(f"Percentage of all trips: {percent_trips:.2f}%")
print(f"Average duration (minutes): {avg_duration:.2f}")


Number of trips matching criteria: 1501
Percentage of all trips: 1.50%
Average duration (minutes): 40.37


#### Task 3d: Explicit Indexing Practice

Practice using `loc[]` and `iloc[]` for different selection tasks:

1. Use `iloc[]` to select the first 10 trips, showing only `member_casual`, `rideable_type`, and `trip_duration_min` columns
2. Use `loc[]` to select trips from October 15-17 (use date strings `'2024-10-15':'2024-10-17'`), showing the same three columns
3. Count how many trips occurred during this date range

Note: When using `iloc[]`, remember it's position-based (0-indexed). When using `loc[]` with the datetime index, you can slice using date strings.

##### Your Code

In [11]:
# (1) iloc[] to select the first 10 trips
display(rides.iloc[:10][['member_casual', 'rideable_type', 'trip_duration_min']])

# (2️) loc[] to select trips between October 15–17 
oct_trips = rides.loc['2024-10-15':'2024-10-17', ['member_casual', 'rideable_type', 'trip_duration_min']]
display(oct_trips)

# how many trips occurred during that date range
trip_count = len(oct_trips)
print(f"Number of trips between October 15–17: {trip_count}")


member_casual  rideable_type  trip_duration_min
started_at                                                             
2024-09-30 23:12:01.622        casual  electric_bike          67.984200
2024-09-30 23:19:25.409        casual  electric_bike          82.742067
2024-09-30 23:32:24.672        member   classic_bike          50.899583
2024-09-30 23:42:11.207        casual   classic_bike          28.093733
2024-09-30 23:49:25.380        member  electric_bike          17.034933
2024-09-30 23:49:40.016        member  electric_bike          13.009367
2024-10-01 00:00:53.414        member   classic_bike           2.598817
2024-10-01 00:05:44.954        member  electric_bike           0.013433
2024-10-01 00:06:12.035        member  electric_bike          10.472933
2024-10-01 00:10:03.646        member   classic_bike           7.825683

member_casual  rideable_type  trip_duration_min
started_at                                                             
2024-10-15 00:00:12.781        casual  electric_bike           2.804233
2024-10-15 00:01:20.517        member  electric_bike          11.330867
2024-10-15 00:05:24.811        member  electric_bike           1.868850
2024-10-15 00:05:52.984        member  electric_bike           2.705083
2024-10-15 00:06:18.819        member  electric_bike           1.600867
...                               ...            ...                ...
2024-10-17 23:45:48.739        member  electric_bike          14.703100
2024-10-17 23:47:35.040        member  electric_bike          13.049867
2024-10-17 23:55:34.112        member   classic_bike           3.795400
2024-10-17 23:56:14.464        member  electric_bike          11.937217
2024-10-17 23:59:38.103        casual   classic_bike           5.266433

[7235 rows x 3 columns]

Number of trips between October 15–17: 7235


#### Interpretation

Reflect on this problem and answer (briefly / concisely) the following questions:

1. `isin()` advantages: Compare using `isin(['A', 'B', 'C'])` versus `(col == 'A') | (col == 'B') | (col == 'C')`. Beyond readability, what practical advantage does `isin()` provide when filtering for many values (e.g., 20+ stations)?
2. Conditional assignment order: In Task 3b, why did we initialize all values to "Midday" before assigning rush hour categories? What would go wrong if you assigned categories in a different order, or didn't set a default?
3. `query()` vs boolean indexing: The `query()` method in Task 3c could have been written with boolean indexing instead. When would you choose `query()` over boolean indexing? When might boolean indexing be preferable despite being more verbose?

##### Your Answers

***
#### isin() advantages
isin() makes things a lot cleaner and simpler, specifically when you’re testing for many values. Instead of writing something long like `(col == 'A') | (col == 'B') | (col == 'C')`, I can just use `col.isin(['A', 'B', 'C'])`. It’s easier to update, read, and maintain later. It’s also faster because Panda handles it internally using efficient set lookups. This is helpful when filtering large datasets with lots of possible categories, like dozens of stations or IDs.In brief, it improves scalability, readability, and performance.

#### Conditional assignment order
I started by setting everything to "Midday" first so that every row has a default value. After that, I used .loc[] to update the rows that fall into "Morning Rush" and "Evening Rush". If I didn’t set a default first, I’d end up with missing (NaN) values in the rows that don’t match either condition. And if I did it in the wrong order, like putting "Midday" last, it would overwrite the rush-hour labels I already set. So doing it this way keeps everything consistent and avoids mistakes.

#### query() vs boolean indexing
I think `query()` is really nice when you just want to write simple, readable filters; it feels more natural, like writing a sentence. It’s especially good for quick exploration or when you’re combining a few conditions with `and` or `or`. But Boolean indexing gives me more flexibility when I need to use string or datetime functions, or when I’m building masks programmatically. So I’d say query() is great for readability, while boolean indexing is better when I need more control or I’m doing something more complex.
***

#### Follow-Up (Graduate Students Only)

Pandas supports a variety of indexing paradigms, including bracket notation (`df['col']`), label-based indexing (`loc[]`), and position-based indexing (`iloc[]`). The lecture recommended using bracket notation only for columns, and loc/iloc for everything else. Explain the rationale: why is this approach better than using bracket notation for everything, even though `df[0:5]` technically works for row slicing?

##### Your Answers

***
Bracket notation, like `df['col']`, is fine for column selection but not for row selection, where it gets error-prone and non-intuitive. This is why the lecture recommends only using bracket notation with column labels, and `.loc[]` or `.iloc[]` for everything else.

The main reason is that `.loc[]` and `.iloc[]` make your intention clear. `.loc[]` is for label-based indexing (using row or column names), while .iloc[] is for position-based indexing (using integer positions). On the other hand, `df[0:5]` can behave inconsistently depending on the situation, and it’s easy to mix up what it’s actually returning. By sticking with `.loc[]` and .iloc[], the code becomes much more readable, predictable, and less likely to break if the index changes or becomes non-numeric.

Ultimately, using .loc[] and .iloc[] clarifies code through reducing ambiguity, especially when working within larger or more complex DataFrames.
***

### Problem 4: Temporal Analysis and Export

Time-based patterns are crucial for understanding bike share usage. In this problem, you'll analyze when trips occur, how usage differs between rider types, and export filtered results. You'll use the datetime index you set in Problem 2 and the derived columns from Problems 2-3.

#### Task 4a: Identify Temporal Patterns

Use the datetime index to extract temporal components and identify usage patterns:

1. Extract the *hour* from the index and use `value_counts()` to find the most popular hour for trips. Report the peak hour and how many trips occurred during that hour.
2. Extract the *day name* from the index and use `value_counts()` to find the busiest day of the week. Report the day and number of trips.
3. Sort the results from step 2 to show days in order from Monday to Sunday (not by trip count). Use `sort_index()`.

Hint: Use `.dt.hour` and `.dt.day_name()` on the datetime index.

##### Your Code

In [12]:
# Extracting hours
hour_counts = rides.index.hour.value_counts().sort_index()

# Finding the most popular hour and number of trips
peak_hour = hour_counts.idxmax()
peak_trips = hour_counts.max()

print(f"Most popular hour for trips: {peak_hour}:00 with {peak_trips} trips")
display(hour_counts) # Verification

# Extracting the day name from the datetime index and count trips per day
day_counts = rides.index.day_name().value_counts()

# Reorder days from Monday to Sunday
ordered_days = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
day_counts = day_counts.reindex(ordered_days)

# Find the busiest day
busiest_day = day_counts.idxmax()
busiest_trips = day_counts.max()

print(f"\nBusiest day of the week: {busiest_day} with {busiest_trips} trips")
display(day_counts)


Most popular hour for trips: 17:00 with 10574 trips


started_at
0      1121
1       719
2       429
3       252
4       226
5       749
6      2281
7      4776
8      6391
9      4703
10     4498
11     5070
12     5818
13     6131
14     6392
15     7541
16     9705
17    10574
18     7939
19     4885
20     3461
21     2725
22     2107
23     1507
Name: count, dtype: int64


Busiest day of the week: Wednesday with 16513 trips


started_at
Monday       11531
Tuesday      14970
Wednesday    16513
Thursday     16080
Friday       13691
Saturday     14427
Sunday       12788
Name: count, dtype: int64

#### Task 4b: Compare Groups with groupby()

Use `groupby()` (introduced in 07a) to compare trip characteristics across different groups:

1. Calculate the average trip duration by rider type (`member_casual`). Which group takes longer trips on average?
2. Calculate the average trip duration by bike type (`rideable_type`). Which bike type has the longest average trip?
3. Count the number of trips by rider type using `groupby()` with `.size()`. Compare this with using `value_counts()` on the `member_casual` column - do they give the same result?

Note: Use single-key groupby only (one column at a time).

##### Your Code

In [13]:
# Average trip duration by rider type
avg_duration_rider = rides.groupby('member_casual')['trip_duration_min'].mean()
print("Average trip duration by rider type:")
display(avg_duration_rider)

#  Identify which rider group takes longer trips on average
longest_group = avg_duration_rider.idxmax()
print(f"\nThe group that takes longer trips on average: {longest_group}")

# Average trip duration by bike type
avg_duration_bike = rides.groupby('rideable_type')['trip_duration_min'].mean()
print("\nAverage trip duration by bike type:")
display(avg_duration_bike)

# Identify which bike type has the longest average trip
longest_bike = avg_duration_bike.idxmax()
print(f"\nThe bike type with the longest average trip: {longest_bike}")

# Counting number of trips by rider type 
groupby_counts = rides.groupby('member_casual').size()
value_counts = rides['member_casual'].value_counts()

print("\nTrip counts using groupby():")
display(groupby_counts)

print("Trip counts using value_counts():")
display(value_counts)

# Compare the two
if groupby_counts.equals(value_counts):
    print("\nBoth methods give the same result.")
else:
    print("\nResults differ — They are the same but order differ.")


Average trip duration by rider type:


member_casual
casual    23.978046
member    11.984493
Name: trip_duration_min, dtype: float64


The group that takes longer trips on average: casual

Average trip duration by bike type:


rideable_type
classic_bike     20.337410
electric_bike    12.033618
Name: trip_duration_min, dtype: float64


The bike type with the longest average trip: classic_bike

Trip counts using groupby():


member_casual
casual    34686
member    65314
dtype: int64

Trip counts using value_counts():


member_casual
member    65314
casual    34686
Name: count, dtype: Int64


Results differ — They are the same but order differ.


#### Task 4c: Filter, Sample, and Export

Create a filtered dataset for weekend electric bike trips and export it:

The provided code once again uses Path to create an `output` directory and constructs the full file path as `output/weekend_electric_trips.csv`. Use the `output_file` variable when calling `.to_csv()`.

1. Filter for trips where `is_weekend == True` and `rideable_type == 'electric_bike'`
2. Use `iloc[]` to select the first 1000 trips from this filtered dataset
3. Use `reset_index()` to convert the datetime index back to a column (so it's included in the export)
4. Export to CSV with filename `weekend_electric_trips.csv`, including only these columns: `started_at`, `ended_at`, `member_casual`, `trip_duration_min`, `time_of_day`
5. Use `index=False` to avoid writing the default numeric index to the file

After exporting, report how many total weekend electric bike trips existed before sampling to 1000.

##### Your Code

In [14]:
# do not modify this setup code
from pathlib import Path

output_dir = Path('output')
output_dir.mkdir(exist_ok=True)
output_file = output_dir / 'weekend_electric_trips.csv'

# Filter for weekend electric bike trips
weekend_electric = rides[(rides['is_weekend'] == True) & (rides['rideable_type'] == 'electric_bike')]

# the total number of weekend electric bike trips before sampling
total_trips = len(weekend_electric)
print(f"Total weekend electric bike trips before sampling: {total_trips}")

# Select the first 1000 trips using iloc
sampled_trips = weekend_electric.iloc[:1000]

# Reset index to include the datetime index as a regular column
sampled_trips = sampled_trips.reset_index()

# Export selected columns to CSV
columns_to_export = ['started_at', 'ended_at', 'member_casual', 'trip_duration_min', 'time_of_day']
sampled_trips[columns_to_export].to_csv(output_file, index=False)

print(f"\nExported {len(sampled_trips)} trips to {output_file}")


Total weekend electric bike trips before sampling: 13026

Exported 1000 trips to output\weekend_electric_trips.csv


#### Interpretation

Reflect on this problem and answer the following questions:

1. `groupby() conceptual model`: Explain in your own words what `groupby()` does. Use the phrase "split-apply-combine" in your explanation and describe what happens at each stage.
2. `value_counts()` vs `groupby()`: In Task 4b.3, you compared two approaches for counting trips by rider type. When would you use `value_counts()` versus `groupby().size()`? Is there a situation where only one of them would work?
3. Index management for export: In Task 4c, why did we use `reset_index()` before exporting? What would happen if you exported with the datetime index still in place and used `index=False`?

##### Your Answers

***
#### groupby() conceptual mode
The groupby function is usually accomplished in three steps, sometimes referred to as split-apply-combine: splitting the data into groups based on the values in one or more column(s), applying a function (for example. mean, count) to each group, and returning a new DataFrame object that contains the results of all the functions. This is used when it is desired to summarize the data, or look at a sub-section without filtering out each data point.

#### value_counts() vs groupby()
Both of these can count how many records are in each category, but they’re useful in slightly different ways. I’d use value_counts() when I just want a quick count from a single column; it’s simple and fast. But if I need to count across multiple columns or do other calculations at the same time, groupby().size() is the better choice. For example, if I wanted to count trips by both rider type and bike type, groupby() would be the only one that works.

#### Index management for export
I used reset_index() here before exporting the CSV so that the datetime index (started_at) would get converted back to a regular old column and therefore saved to the outputted CSV. If I hadn't reset the index and used index=False, then those datetime values wouldn't have been saved to the outputted CSV at all. Resetting the index just ensures that all the information, such as the timestamps, is included in the export
***

#### Follow-Up (Graduate Students Only)

Compare `CSV` and _pickle_ formats for data storage and retrieval.

Pickle is Python's built-in serialization format that saves Python objects exactly as they exist in memory, preserving all data types, structures, and metadata. Unlike CSV (which converts everything to text), pickle is binary (not human readable) and maintains the complete state of your DataFrame. Also, pickle files only work in Python, while CSV is universal. Read more in the [Pandas documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_pickle.html).

The code below investigates an interesting pattern: Do riders take longer trips from scenic lakefront stations even during rush hours? This could indicate tourists or recreational riders using these popular locations for leisure trips during typical commute times. The analysis filters for trips over 15 minutes that started from lakefront stations during morning (7-9am) or evening (4-6pm) rush hours, sorted by duration to see the longest trips first.

Run the code below, then answer the interpretation questions:

In [16]:
import os

# the following lines were commented out since they were run in 4c
# from pathlib import Path
# output_dir = Path('output')

csv_file = output_dir / 'lakefront_rush_trips.csv'
pickle_file = output_dir / 'lakefront_rush_trips.pkl'

# Filter for interesting pattern: Long trips (>15 min) during rush hours 
# from lakefront stations, sorted by duration
lakefront_rush = (rides

 .loc[(rides.index.hour.isin([7, 8, 9, 16, 17, 18])) &

 (rides['start_station_name'].str.contains('Lake Shore|Lakefront',

 case=False,

 na=False)) &

 (rides['trip_duration_min'] > 15)]

 .sort_values('trip_duration_min', ascending=False)

 .head(1000)

 .reset_index()

 [['started_at', 'ended_at', 'start_station_name', 'end_station_name',

 'member_casual', 'rideable_type', 'trip_duration_min']]

)


print(f"Found {len(lakefront_rush)} long rush-hour trips from lakefront stations")

# Export to both formats
lakefront_rush.to_csv(csv_file, index=False)
lakefront_rush.to_pickle(pickle_file)

# Compare file sizes
csv_size = os.path.getsize(csv_file) / 1024  # Convert to KB
pickle_size = os.path.getsize(pickle_file) / 1024
print(f"\nCSV file size: {csv_size:.2f} KB")
print(f"Pickle file size: {pickle_size:.2f} KB")
print(f"Size difference: {abs(csv_size - pickle_size):.2f} KB")

# Compare load times
print("\nLoad time comparison:")
print("CSV:")
%timeit pd.read_csv(csv_file)
print("\nPickle:")
%timeit pd.read_pickle(pickle_file)

# Check data type preservation
# Note: CSV load without parse_dates loses datetime types
csv_loaded = pd.read_csv(csv_file)
pickle_loaded = pd.read_pickle(pickle_file)

print("\nData types from CSV (without parse_dates):")
print(csv_loaded.dtypes)
print("\nData types from Pickle:")
print(pickle_loaded.dtypes)

Found 498 long rush-hour trips from lakefront stations

CSV file size: 70.12 KB
Pickle file size: 25.73 KB
Size difference: 44.40 KB

Load time comparison:
CSV:
2.14 ms ± 256 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

Pickle:
431 μs ± 46.3 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

Data types from CSV (without parse_dates):
started_at             object
ended_at               object
start_station_name     object
end_station_name       object
member_casual          object
rideable_type          object
trip_duration_min     float64
dtype: object

Data types from Pickle:
started_at            datetime64[ns]
ended_at              datetime64[ns]
start_station_name    string[python]
end_station_name      string[python]
member_casual         string[python]
rideable_type         string[python]
trip_duration_min            float64
dtype: object


After running the code, answer these questions:

1. Method chaining: The analysis uses method chaining with a specific formatting pattern:

   ```python
   result = (df
       .method1()
       .method2()
       .method3()
   )
   ```

   This wraps the entire chain in parentheses, allowing each method to appear on its own line without backslashes. Discuss why this makes formatting more readable, how it makes debugging easier, how it relates to seeing changes in the code with git diff, and what downsides heavy chaining might have.
3. Data types: Compare the dtypes from CSV versus pickle. What types were preserved by pickle that were lost in CSV? Why is this preservation significant for subsequent analysis?
4. Trade-offs: Given your observations about size, speed, and type preservation, when would you choose pickle over CSV for your work? When would CSV still be the better choice despite pickle's advantages?


***
#### Method chaining
The use of method chaining can improve the readability of code. Method chaining makes the program easier to read because every call is a new line in the code. This obviously shows what actions happen on a list. Sorts, filters, and groups are chained.

The format lets people debug or version-control the code by using Git, and makes it obvious any change in something that is being done because each operation exists on its own line. This is useful when something does not work because I can comment it out or test a step in the chain without rewriting the code.

However, the downside of heavy method chaining is that if the chain gets too long, it becomes harder to debug or inspect intermediate results. It can also make the code less clear and difficult for someone still learning the language to follow along, since all the transformations are packed into one expression. For complex tasks, breaking the process into smaller steps can sometimes be more readable and easier to maintain. For complex tasks, splitting the process into smaller steps can sometimes be more readable and easier to maintain.

#### Data types 
I looked into the CSV and into the pickle files. I noticed that the data types were preserved in the pickle version of the data. The data types were either modified or lost in the CSV version. For example, the `started_at` and `ended_at columns` stayed as proper datetime objects `(datetime64[ns])` in the pickle file, but when loaded from CSV, they became just plain text (object type).

The difference matters because using the right datatypes eases analysis much: with the pickle file, I can use datetime methods like .dt.hour and .dt.day_name() right away. If I do use the CSV version of the data, I would have to parse the dates again in turn which is slower and more error-prone.

Overall, pickle preserves the structure and meaning of the data, which is really important for any further analysis or calculations.

#### Trade-offs
Pickle is generally going to be faster to load, going to take up less file space, and going to preserve all of the data types exactly as they were. So if I'm working in pure Python, and I'm going to be saving and reloading data on an iterative basis, I use pickle, particularly if datetime and category are important. 

However, the CSV format is still the better option when I want to share my data with others and open it in other software, such as Excel or R. CSV is a text-based format and human-readable, while pickle files only work in Python. Even though pickle is more efficient, CSV is much more portable and safer to use for collaboration or long-term storage. In conclusion, pickle is great for performance and accuracy, while CSV is better for compatibility and sharing.

#### What did I change from the original code with this fix? In your opinion, why didn't the original code work and why does this one work as intended? 
In the fixed version, all of the filtering was put into one boolean expression in a single `.loc()` statement rather than multiple `.loc()` statements. In the original code, every `.loc()`  used a mask based on the entire dataset, but was applied to a DataFrame that had already been filtered in the previous step, which would result in a mismatch between the filtered DataFrame and the mask length.

The fixed version works because it combines all the conditions into one DataFrame, and the mask remains aligned with the data and allows missing data to be processed without causing an exception by using `na=False` in `str.contains()`.
***

## Reflection

Address the following questions in a markdown cell:

1. NumPy vs Pandas
   - What was the biggest conceptual shift moving from NumPy arrays to Pandas DataFrames?
   - Which Pandas concept was most challenging: indexing (loc/iloc), missing data, datetime operations, or method chaining? How did you work through it?
2. Real Data Experience
   - How did working with real CSV data (with missing values, datetime strings, etc.) differ from hw2b's synthetic NumPy arrays?
   - Based on this assignment, what makes Pandas well-suited for data analysis compared to pure NumPy?
3. Learning & Application
   - Which new skill from this assignment will be most useful for your own data work?
   - On a scale of 1-10, how prepared do you feel to use Pandas for your own projects? What would increase that score?
4. Feedback
   - Time spent: ___ hours (breakdown optional)
   - Most helpful part of the assignment: ___
   - One specific improvement suggestion: ___

### Your Answers

***
#### Numpy vs Pandas 
For me, the biggest conceptual change was using labeled data instead of numeric arrays. NumPy involves arrays of numbers and positional indexing, but Pandas uses a data structure associated with labels. Every column in Pandas has its own label along with multiple data types, making it more like working with a database than a matrix.

The hardest concept in pandas, for me, really was the indexing, especially making sense of loc and iloc. In the beginning, the concept of when to use label-based vs. position-based selection, but after a time practing with real data and seeing how loc works with datetime indexes, it started to make more sense.

#### Real Data Experience
Working with real CSV data was different from working with clean NumPy arrays in that the CSV files we received had missing data, strings, and datetime columns that we needed to transform. I also learned that real-world data always needed cleaning, and Pandas made the process easier than NumPy did.

Pandas is well-suited for data analysis because it provides tools for handling missing data, grouping, filtering, and time-based analysis, all things which are hard to do in NumPy. It also integrates naturally with file formats like CSV and makes exploring data faster and intuitive.

#### Learning & Application
The most useful new skill I learned from this assignment is using `groupby()` for data `aggregation` and `comparison`. Before this, I mostly analyzed data column by column, but groupby() showed me how to quickly summarize large datasets and compare different categories, like member types or bike types, in just a few lines of code

I’d rate my confidence around 7. I feel pretty comfortable using Pandas for most of my projects now, but I’d improve even more by practicing more advanced topics like merging datasets and using groupby with multiple keys, not only one key at a time.

#### Feedback
Time spent: 8 hours (including experimenting and debugging)  

Most helpful part of the assignment: The section on datetime operations and indexing, which helped me understand how powerful time-based data handling is in Pandas.  

One specific improvement suggestion: It would be nice to have a small cheat sheet summarizing common Pandas methods and their arguments to make it easier to reference while coding or on the exam. 

## Thank you
***